File: src/colibri/query.py
"""Query engine for semantic search over content sources.

Uses direct LanceDB SDK and Ollama HTTP API — no framework dependency.
"""

import logging
import math
import re
from collections import Counter
from pathlib import Path

import lancedb

from colibri.config import (
    EMBEDDING_MODEL,
    LANCEDB_DIR,
    OLLAMA_BASE_URL,
    SIMILARITY_THRESHOLD,
    SOURCES,
    TOP_K,
)
from colibri.embedding import embed_texts
from colibri.index_meta import SCHEMA_VERSION, read_index_meta
from colibri.source_factory import compute_nested_exclusions, create_source_for_profile
from colibri.sources import ContentSource, SourceDocument

logger = logging.getLogger(__name__)


class SearchEngine:
    """Search engine for content sources with semantic retrieval."""

    def __init__(
        self,
        table_name: str = "chunks",
        source: ContentSource | None = None,
    ):
        """Initialize the search engine.

        Args:
            table_name: LanceDB table name
            source: Content source (if provided, used as primary source)
        """
        # Build all sources from SOURCES config
        nested_map = compute_nested_exclusions(SOURCES)
        self._all_sources: list[ContentSource] = []

        if source is not None:
            self._all_sources.append(source)

        for profile in SOURCES:
            src = create_source_for_profile(profile, nested_map)
            # Avoid adding duplicates when a custom source was passed
            if source is not None and src.root_path == source.root_path:
                continue
            self._all_sources.append(src)

        self.source = source or (self._all_sources[0] if self._all_sources else None)
        self.index_path = LANCEDB_DIR
        self.table_name = table_name

        # Check schema version before opening the table
        meta = read_index_meta(self.index_path)
        stored_version = meta.get("schema_version", 0)
        if stored_version != SCHEMA_VERSION:
            msg = (
                f"Index schema outdated (v{stored_version}, need v{SCHEMA_VERSION}). "
                f"Run `colibri index --force` to rebuild."
            )
            raise RuntimeError(msg)

        # Connect to LanceDB
        self._db = lancedb.connect(str(self.index_path))
        self._table = self._db.open_table(table_name)

    def search(
        self,
        query: str,
        folder: str | None = None,
        doc_type: str | None = None,
        limit: int = 5,
    ) -> list[dict]:
        """Search library content with optional folder or doc_type filter.

        Args:
            query: The search query
            folder: Optional folder name to filter results (e.g., "Books")
            doc_type: Optional document type filter (e.g., "book")
            limit: Maximum number of results to return

        Returns:
            List of search results with text, metadata, and scores
        """
        # Embed the query
        query_vector = embed_texts([query], model=EMBEDDING_MODEL, base_url=OLLAMA_BASE_URL)[0]

        # Build search query
        search_query = self._table.search(query_vector).limit(TOP_K)

        # Apply folder pre-filter (use startswith to handle nested folder paths)
        if folder:
            escaped_folder = folder.replace("'", "''")
            search_query = search_query.where(
                f"folder = '{escaped_folder}' OR source_file LIKE '{escaped_folder}/%'"
            )

        # Apply doc_type filter
        if doc_type:
            escaped_type = doc_type.replace("'", "''")
            search_query = search_query.where(f"doc_type = '{escaped_type}'")

        raw_results = search_query.to_list()

        results = []
        for row in raw_results:
            # Convert L2 distance to similarity score (0-1 range)
            distance = row.get("_distance", 0.0)
            score = math.exp(-distance)

            if score < SIMILARITY_THRESHOLD:
                continue

            source_file = row.get("source_file", "")
            results.append(
                {
                    "text": row.get("text", ""),
                    "file": source_file,
                    "title": row.get("title", Path(source_file).stem),
                    "type": row.get("doc_type", "note"),
                    "folder": row.get("folder", ""),
                    "score": round(score, 4),
                }
            )

            if len(results) >= limit:
                break

        return results

    def search_books(self, query: str, limit: int = 5) -> list[dict]:
        """Search only documents with doc_type='book'."""
        return self.search(query, doc_type="book", limit=limit)

    def search_library(self, query: str, limit: int = 5) -> list[dict]:
        """Search the entire library."""
        return self.search(query, limit=limit)

    def _resolve_document(self, note_path: str) -> SourceDocument | None:
        """Try to read a document from any known source.

        Args:
            note_path: Path relative to a source root

        Returns:
            SourceDocument if found in any source, None otherwise
        """
        path = Path(note_path)
        for source in self._all_sources:
            try:
                return source.read_document(path)
            except (FileNotFoundError, ValueError):
                continue
        return None

    def get_note(self, note_path: str) -> dict | None:
        """Get a specific note by its path.

        Searches across all configured sources (primary library and
        external sources).

        Args:
            note_path: Path relative to source root (e.g., "Books/Clean Architecture.md")

        Returns:
            Dict with note content and metadata, or None if not found
        """
        doc = self._resolve_document(note_path)
        if doc is None:
            return None
        return {
            "path": note_path,
            "title": doc.title,
            "metadata": doc.metadata,
            "content": doc.content,
        }

    def get_linked_notes(self, note_path: str) -> list[str]:
        """Extract internal links from a note.

        Supports both [[wiki links]] and standard [text](path.md) links.

        Args:
            note_path: Path relative to source root

        Returns:
            List of linked note names/paths
        """
        doc = self._resolve_document(note_path)
        if doc is None:
            return []

        # Extract wiki links: [[link]] and [[link|alias]]
        wiki_links = re.findall(r"\[\[([^\]|]+)(?:\|[^\]]+)?\]\]", doc.content)
        # Extract markdown links to .md files: [text](path.md)
        md_links = re.findall(r"\[([^\]]+)\]\(([^)]+\.md)\)", doc.content)
        return wiki_links + [path for _, path in md_links]

    def list_books(self) -> list[dict]:
        """List all indexed books with metadata from frontmatter.

        Returns title, chunk count, and — when the source file is available —
        author, language, tags, and translation info.
        """
        try:
            df = self._table.to_pandas()

            books_df = df[df["doc_type"] == "book"]
            if books_df.empty:
                return []

            book_stats = books_df.groupby("title").agg(
                chunks=("title", "size"),
                source_file=("source_file", "first"),
            ).reset_index()

            results: list[dict] = []
            for _, row in book_stats.iterrows():
                entry: dict = {
                    "title": row["title"],
                    "chunks": int(row["chunks"]),
                    "file": row["source_file"],
                }

                # Enrich with frontmatter metadata when available
                try:
                    doc = self._resolve_document(row["source_file"])
                    if doc is None:
                        raise FileNotFoundError(row["source_file"])
                    meta = doc.metadata
                    if meta.get("author"):
                        entry["author"] = meta["author"]
                    if meta.get("language"):
                        entry["language"] = meta["language"]
                    if doc.tags:
                        entry["tags"] = doc.tags
                    if meta.get("original_title"):
                        entry["original_title"] = meta["original_title"]
                    if meta.get("translated_from"):
                        entry["translated_from"] = meta["translated_from"]
                except (FileNotFoundError, ValueError):
                    logger.debug("Source file unavailable for %s", row["title"])

                results.append(entry)

            return results

        except Exception as e:
            logger.warning("list_books failed: %s", e)
            return []

    def get_book_outline(self, file_path: str) -> list[dict] | None:
        """Get the heading structure (table of contents) from a document.

        Searches across all configured sources.

        Args:
            file_path: Path relative to source root (e.g. "Books/Clean Architecture.md")

        Returns:
            List of {level, text} dicts for each heading, or None if file not found.
        """
        doc = self._resolve_document(file_path)
        if doc is None:
            return None

        headings: list[dict] = []
        for match in re.finditer(r"^(#{1,3})\s+(.+)$", doc.content, re.MULTILINE):
            headings.append({
                "level": len(match.group(1)),
                "text": match.group(2).strip(),
            })
        return headings

    def browse_topics(self, folder: str | None = None) -> list[dict]:
        """List all topics (tags) with document counts.

        Deduplicates by source file so a multi-chunk document
        only counts once per tag.

        Args:
            folder: Optional folder filter (e.g. "Books")

        Returns:
            Sorted list of {tag, document_count} dicts.
        """
        try:
            df = self._table.to_pandas()

            if folder:
                df = df[df["folder"] == folder]

            # Deduplicate: one row per unique source_file
            unique_docs = df.drop_duplicates(subset=["source_file"])[
                ["source_file", "tags"]
            ]

            tag_counter: Counter[str] = Counter()
            for tags_str in unique_docs["tags"]:
                if not tags_str:
                    continue
                for tag in str(tags_str).split(","):
                    tag = tag.strip()
                    if tag:
                        tag_counter[tag] += 1

            return [
                {"tag": tag, "document_count": count}
                for tag, count in tag_counter.most_common()
            ]

        except Exception as e:
            logger.warning("browse_topics failed: %s", e)
            return []


# Singleton instance
_engine: SearchEngine | None = None


def get_engine() -> SearchEngine:
    """Get or create the search engine singleton."""
    global _engine
    if _engine is None:
        _engine = SearchEngine()
    return _engine

File: colibri-rs/src/query.rs
//! Query engine for semantic search over content sources.
//!
//! Mirrors the Python `query.py` module. Uses LanceDB vector search
//! with L2 distance converted to similarity score via `exp(-distance)`.

use std::collections::{HashMap, HashSet};
use std::path::Path;

use arrow_array::{Float32Array, RecordBatch, StringArray};
use futures::TryStreamExt;
use lancedb::query::{ExecutableQuery, QueryBase, Select};
use serde::Serialize;

use crate::config::{AppConfig, SCHEMA_VERSION};
use crate::embedding::embed_texts;
use crate::error::ColibriError;
use crate::index_meta::read_index_meta;

/// A single search result.
#[derive(Debug, Clone, Serialize)]
pub struct SearchResult {
    pub text: String,
    pub file: String,
    pub title: String,
    #[serde(rename = "type")]
    pub doc_type: String,
    pub folder: String,
    pub score: f64,
}

/// Book entry from the index.
#[derive(Debug, Clone, Serialize)]
pub struct BookEntry {
    pub title: String,
    pub chunks: usize,
    pub file: String,
}

/// Topic entry with document count.
#[derive(Debug, Clone, Serialize)]
pub struct TopicEntry {
    pub tag: String,
    pub document_count: usize,
}

/// Search engine backed by LanceDB.
pub struct SearchEngine {
    table: lancedb::Table,
    config: AppConfig,
}

impl SearchEngine {
    /// Create a new search engine, verifying schema version.
    pub async fn new(config: &AppConfig) -> Result<Self, ColibriError> {
        // Check schema version
        let meta = read_index_meta(&config.lancedb_dir)?;
        let stored_version = meta
            .get("schema_version")
            .and_then(|v| v.as_u64())
            .unwrap_or(0) as u32;

        if stored_version != SCHEMA_VERSION {
            return Err(ColibriError::Query(format!(
                "Index schema outdated (v{stored_version}, need v{SCHEMA_VERSION}). \
                 Run `colibri index --force` to rebuild."
            )));
        }

        let db = lancedb::connect(config.lancedb_dir.to_string_lossy().as_ref())
            .execute()
            .await?;

        let table = db.open_table("chunks").execute().await.map_err(|e| {
            ColibriError::Query(format!(
                "Failed to open 'chunks' table. Run `colibri index` first. ({e})"
            ))
        })?;

        Ok(Self {
            table,
            config: config.clone(),
        })
    }

    /// Semantic search with optional folder/doc_type filters.
    pub async fn search(
        &self,
        query: &str,
        folder: Option<&str>,
        doc_type: Option<&str>,
        limit: usize,
    ) -> Result<Vec<SearchResult>, ColibriError> {
        // Embed the query
        let query_vector = embed_texts(
            &[query.to_string()],
            &self.config.embedding_model,
            &self.config.ollama_base_url,
        )
        .await?;

        if query_vector.is_empty() {
            return Err(ColibriError::Embedding(
                "Query embedding returned no results".into(),
            ));
        }

        // Build search
        let mut search = self
            .table
            .vector_search(query_vector[0].clone())
            .map_err(|e| ColibriError::Query(format!("Vector search failed: {e}")))?
            .limit(self.config.top_k);

        // Apply filters
        if let Some(folder) = folder {
            let escaped = folder.replace('\'', "''");
            search = search.only_if(format!(
                "folder = '{escaped}' OR source_file LIKE '{escaped}/%'"
            ));
        }
        if let Some(dt) = doc_type {
            let escaped = dt.replace('\'', "''");
            search = search.only_if(format!("doc_type = '{escaped}'"));
        }

        let raw_results = search
            .execute()
            .await
            .map_err(|e| ColibriError::Query(format!("Search execution failed: {e}")))?;

        // Collect results from the stream
        let batches: Vec<RecordBatch> = raw_results
            .try_collect()
            .await
            .map_err(|e| ColibriError::Query(format!("Failed to collect results: {e}")))?;

        let mut results = Vec::new();

        for batch in &batches {
            let num_rows = batch.num_rows();

            let text_col = batch
                .column_by_name("text")
                .and_then(|c| c.as_any().downcast_ref::<StringArray>());
            let file_col = batch
                .column_by_name("source_file")
                .and_then(|c| c.as_any().downcast_ref::<StringArray>());
            let title_col = batch
                .column_by_name("title")
                .and_then(|c| c.as_any().downcast_ref::<StringArray>());
            let type_col = batch
                .column_by_name("doc_type")
                .and_then(|c| c.as_any().downcast_ref::<StringArray>());
            let folder_col = batch
                .column_by_name("folder")
                .and_then(|c| c.as_any().downcast_ref::<StringArray>());
            let dist_col = batch
                .column_by_name("_distance")
                .and_then(|c| c.as_any().downcast_ref::<Float32Array>());

            for i in 0..num_rows {
                let distance = dist_col.map(|c| c.value(i) as f64).unwrap_or(0.0);
                let score = (-distance).exp();

                if score < self.config.similarity_threshold {
                    continue;
                }

                let source_file = file_col.map(|c| c.value(i)).unwrap_or("");
                let title = title_col.map(|c| c.value(i)).unwrap_or_else(|| {
                    Path::new(source_file)
                        .file_stem()
                        .and_then(|s| s.to_str())
                        .unwrap_or("")
                });

                results.push(SearchResult {
                    text: text_col.map(|c| c.value(i).to_string()).unwrap_or_default(),
                    file: source_file.to_string(),
                    title: title.to_string(),
                    doc_type: type_col
                        .map(|c| c.value(i).to_string())
                        .unwrap_or_else(|| "note".to_string()),
                    folder: folder_col
                        .map(|c| c.value(i).to_string())
                        .unwrap_or_default(),
                    score: (score * 10000.0).round() / 10000.0,
                });

                if results.len() >= limit {
                    return Ok(results);
                }
            }
        }

        Ok(results)
    }

    /// Search only books.
    pub async fn search_books(
        &self,
        query: &str,
        limit: usize,
    ) -> Result<Vec<SearchResult>, ColibriError> {
        self.search(query, None, Some("book"), limit).await
    }

    /// Search the entire library.
    pub async fn search_library(
        &self,
        query: &str,
        limit: usize,
    ) -> Result<Vec<SearchResult>, ColibriError> {
        self.search(query, None, None, limit).await
    }

    /// List all indexed books with metadata.
    pub async fn list_books(&self) -> Result<Vec<BookEntry>, ColibriError> {
        // Query all book rows
        let results = self
            .table
            .query()
            .only_if("doc_type = 'book'")
            .select(Select::columns(&["title", "source_file"]))
            .execute()
            .await
            .map_err(|e| ColibriError::Query(format!("list_books query failed: {e}")))?;

        let batches: Vec<RecordBatch> = results
            .try_collect()
            .await
            .map_err(|e| ColibriError::Query(format!("Failed to collect: {e}")))?;

        // Aggregate by title
        let mut book_map: HashMap<String, (usize, String)> = HashMap::new();

        for batch in &batches {
            let title_col = batch
                .column_by_name("title")
                .and_then(|c| c.as_any().downcast_ref::<StringArray>());
            let file_col = batch
                .column_by_name("source_file")
                .and_then(|c| c.as_any().downcast_ref::<StringArray>());

            for i in 0..batch.num_rows() {
                let title = title_col
                    .map(|c| c.value(i).to_string())
                    .unwrap_or_default();
                let file = file_col.map(|c| c.value(i).to_string()).unwrap_or_default();
                let entry = book_map.entry(title).or_insert((0, file));
                entry.0 += 1;
            }
        }

        let mut books: Vec<BookEntry> = book_map
            .into_iter()
            .map(|(title, (chunks, file))| BookEntry {
                title,
                chunks,
                file,
            })
            .collect();

        books.sort_by(|a, b| a.title.cmp(&b.title));
        Ok(books)
    }

    /// List all topics (tags) with document counts.
    pub async fn browse_topics(
        &self,
        folder: Option<&str>,
    ) -> Result<Vec<TopicEntry>, ColibriError> {
        let query = if let Some(folder) = folder {
            let escaped = folder.replace('\'', "''");
            self.table
                .query()
                .only_if(format!("folder = '{escaped}'"))
                .select(Select::columns(&["source_file", "tags"]))
                .execute()
                .await
        } else {
            self.table
                .query()
                .select(Select::columns(&["source_file", "tags"]))
                .execute()
                .await
        };

        let results =
            query.map_err(|e| ColibriError::Query(format!("browse_topics failed: {e}")))?;

        let batches: Vec<RecordBatch> = results
            .try_collect()
            .await
            .map_err(|e| ColibriError::Query(format!("Failed to collect: {e}")))?;

        // Deduplicate by source_file, then count tags
        let mut seen_files: HashSet<String> = HashSet::new();
        let mut tag_counter: HashMap<String, usize> = HashMap::new();

        for batch in &batches {
            let file_col = batch
                .column_by_name("source_file")
                .and_then(|c| c.as_any().downcast_ref::<StringArray>());
            let tags_col = batch
                .column_by_name("tags")
                .and_then(|c| c.as_any().downcast_ref::<StringArray>());

            for i in 0..batch.num_rows() {
                let file = file_col.map(|c| c.value(i).to_string()).unwrap_or_default();
                if !seen_files.insert(file) {
                    continue; // already counted this file
                }
                let tags_str = tags_col.map(|c| c.value(i).to_string()).unwrap_or_default();
                if tags_str.is_empty() {
                    continue;
                }
                for tag in tags_str.split(',') {
                    let tag = tag.trim();
                    if !tag.is_empty() {
                        *tag_counter.entry(tag.to_string()).or_default() += 1;
                    }
                }
            }
        }

        let mut topics: Vec<TopicEntry> = tag_counter
            .into_iter()
            .map(|(tag, count)| TopicEntry {
                tag,
                document_count: count,
            })
            .collect();

        // Sort by count descending
        topics.sort_by(|a, b| b.document_count.cmp(&a.document_count));
        Ok(topics)
    }
}
